gax/backoff_policy_unittest.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <random>
+
+#include "googletest/include/gtest/gtest.h"
+
+#include "backoff_policy.h"
+
+namespace google {
+namespace gax {
+
+TEST(ExponentialBackoffPolicy, Basic) {
+  ExponentialBackoffPolicy tested(std::chrono::milliseconds(1),
+                                  std::chrono::milliseconds(32));
+
+  for(int i = 0; i < 5; ++i){
+    auto base_delay = tested.CurrentDelay();
+    EXPECT_EQ(base_delay, std::chrono::milliseconds(1<<i));
+    auto delay = tested.OnCompletion();
+    EXPECT_GE(delay, base_delay/2.0);
+    EXPECT_LE(delay, base_delay);
+  }
+
+  EXPECT_EQ(tested.CurrentDelay(), tested.MaximumDelay());
+  // current_delay_range_ saturates to max delay.
+  tested.OnCompletion();
+  EXPECT_EQ(tested.CurrentDelay(), tested.MaximumDelay());
+}
+
+TEST(ExponentialBackoffPolicy, CopyConstruct) {
+  ExponentialBackoffPolicy tested(std::chrono::milliseconds(10),
+                                  std::chrono::milliseconds(320));
+  tested.OnCompletion();
+  ExponentialBackoffPolicy copy(tested);  // Copy starts with fresh backoff delays
+
+  EXPECT_EQ(copy.CurrentDelay(), std::chrono::milliseconds(10));
+  EXPECT_EQ(copy.MaximumDelay(), std::chrono::milliseconds(320));
+}
+
+TEST(ExponentialBackoffPolicy, MoveConstruct) {
+  ExponentialBackoffPolicy tested(std::chrono::milliseconds(10),
+                                  std::chrono::milliseconds(320));
+  tested.OnCompletion();
+  ExponentialBackoffPolicy moved(std::move(tested));  // Starts with fresh backoff delays
+
+  EXPECT_EQ(moved.CurrentDelay(), std::chrono::milliseconds(10));
+  EXPECT_EQ(moved.MaximumDelay(), std::chrono::milliseconds(320));
+}
+
+TEST(ExponentialBackoffPolicy, Clone) {
```

Comment:
Can we test this in the same way as we implement the Basic test, by checking it is between range of values? We know the initial value, don't need CurrentDelay to get it.

Reply:
That's absolutely a thing we could do, but I like having the strong guarantee that the clone is 'fresh'. I can do that via the same static cast/unique_ptr release and then direct access that I used to check that the clone lazily constructs its rng. WDYT?

Reply:
SGTM